RecoTracker/TkSeedGenerator/src/SeedingOTEDProducer.cc

```diff
@@ -0,0 +1,319 @@
+            const TrackingRecHit* hitL1 = mL1.recHit().get();
+
+            //propagate to the L1 and update the TSOS
+            std::pair<bool, TrajectoryStateOnSurface> updatedTSOSL1 =
+                propagateAndUpdate(updatedTSOS.second, *searchingPropagator, *hitL1);
+            if (!updatedTSOSL1.first)
+              continue;
+
+            edm::OwnVector<TrackingRecHit> container;
+            container.push_back(hitL1->clone());
+            container.push_back(hitL2->clone());
+            container.push_back(hitL3.clone());
+
+            //building trajectory inside-out
+            if (searchingPropagator->propagationDirection() == alongMomentum) {
+              buildingPropagator->setPropagationDirection(oppositeToMomentum);
+            } else if (searchingPropagator->propagationDirection() == oppositeToMomentum) {
+              buildingPropagator->setPropagationDirection(alongMomentum);
+            }
+
+            updatedTSOSL1.second.rescaleError(100);
+
+            TrajectoryStateOnSurface updatedTSOSL1_final = theUpdator->update(updatedTSOSL1.second, *hitL1);
+            if
+              UNLIKELY(!updatedTSOSL1_final.isValid()) continue;
+            std::pair<bool, TrajectoryStateOnSurface> updatedTSOSL2_final =
+                propagateAndUpdate(updatedTSOSL1_final, *buildingPropagator, *hitL2);
+            std::pair<bool, TrajectoryStateOnSurface> updatedTSOSL3_final =
+                propagateAndUpdate(updatedTSOSL2_final.second, *buildingPropagator, hitL3);
+            TrajectorySeed ts =
+                createSeed(updatedTSOSL3_final.second, container, hitL3.geographicalId(), *buildingPropagator);
+            result.push_back(ts);
+          }
+        }
+      }
+    }
+  }
+
+  return result;
+}
+
+unsigned int SeedingOTEDProducer::checkLayer(unsigned int iidd) {
+  StripSubdetector strip = StripSubdetector(iidd);
+  unsigned int subid = strip.subdetId();
+  if (subid == StripSubdetector::TIB || subid == StripSubdetector::TOB) {
+    return tkTopo->layer(iidd);
+  }
+  return 0;
+}
+
+std::vector<VectorHit> SeedingOTEDProducer::collectVHsOnLayer(edm::Handle<VectorHitCollectionNew> VHs,
+                                                              unsigned int layerNumber) {
+  const VectorHitCollectionNew& input = *VHs;
+  std::vector<VectorHit> VHsOnLayer;
+  if (!input.empty()) {
+    for (auto DSViter : input) {
+      if (checkLayer(DSViter.id()) == layerNumber) {
+        for (auto vh : DSViter) {
+          VHsOnLayer.push_back(vh);
+        }
+      }
+    }
+  }
+
+  return VHsOnLayer;
+}
+
+void SeedingOTEDProducer::printVHsOnLayer(edm::Handle<VectorHitCollectionNew> VHs, unsigned int layerNumber) {
+  const VectorHitCollectionNew& input = *VHs;
+  if (!input.empty()) {
+    for (auto DSViter : input) {
+      for (auto vh : DSViter) {
+        if (checkLayer(DSViter.id()) == layerNumber)
+          std::cout << " VH in layer " << layerNumber << " >> " << vh << std::endl;
```

Comment:
nicer to let this function take any output stream (then we also make sure cout is never called in production)

Reply:
@kpedro88 I'm not entirely sure what  change you are proposing.

Reply:
example: https://github.com/cms-sw/cmssw/blob/master/DataFormats/HcalDetId/src/HcalDetId.cc
doesn't have to be an `operator<<`, but should take an `std::ostream` object as a parameter and print to that object.
That way, test code (like an analyzer) can pass `std::cout` if it wants, but production code would not use `std::cout` (rather something like LogInfo if necessary in a specific situation).

Reply:
Got, it thanks! I have change it accordingly.